modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/ZoneRebalanceRaftGroupEventsListener.java

```diff
@@ -0,0 +1,615 @@
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.ignite.internal.affinity.AffinityUtils;
+import org.apache.ignite.internal.affinity.Assignment;
+import org.apache.ignite.internal.affinity.Assignments;
+import org.apache.ignite.internal.catalog.CatalogService;
+import org.apache.ignite.internal.distributionzones.DistributionZoneManager;
+import org.apache.ignite.internal.lang.ByteArray;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.metastorage.Entry;
+import org.apache.ignite.internal.metastorage.MetaStorageManager;
+import org.apache.ignite.internal.metastorage.WatchEvent;
+import org.apache.ignite.internal.metastorage.dsl.Condition;
+import org.apache.ignite.internal.metastorage.dsl.Iif;
+import org.apache.ignite.internal.metastorage.dsl.SimpleCondition;
+import org.apache.ignite.internal.metastorage.dsl.Update;
+import org.apache.ignite.internal.raft.PeersAndLearners;
+import org.apache.ignite.internal.raft.RaftError;
+import org.apache.ignite.internal.raft.RaftGroupEventsListener;
+import org.apache.ignite.internal.raft.Status;
+import org.apache.ignite.internal.replicator.ZonePartitionId;
+import org.apache.ignite.internal.util.ByteUtils;
+import org.apache.ignite.internal.util.IgniteSpinBusyLock;
+
+/**
+ * Listener for the raft group events, which must provide correct error handling of rebalance process and start new rebalance after the
+ * current one finished.
+ */
+public class ZoneRebalanceRaftGroupEventsListener implements RaftGroupEventsListener {
+    /** Ignite logger. */
+    private static final IgniteLogger LOG = Loggers.forClass(ZoneRebalanceRaftGroupEventsListener.class);
+
+    /** Number of retrying of the current rebalance in case of errors. */
+    private static final int REBALANCE_RETRY_THRESHOLD = 10;
+
+    /** Delay between unsuccessful trial of a rebalance and a new trial, ms. */
+    private static final int REBALANCE_RETRY_DELAY_MS = 200;
+
+    /** Success code for the MetaStorage switch append assignments change. */
+    private static final int SWITCH_APPEND_SUCCESS = 1;
+
+    /** Success code for the MetaStorage switch reduce assignments change. */
+    private static final int SWITCH_REDUCE_SUCCESS = 2;
+
+    /** Success code for the MetaStorage pending rebalance change. */
+    private static final int SCHEDULE_PENDING_REBALANCE_SUCCESS = 3;
+
+    /** Success code for the MetaStorage stable assignments change. */
+    private static final int FINISH_REBALANCE_SUCCESS = 4;
```

Comment:
What's the difference between ZoneRebalanceRaftGroupEventsListener and RebalanceRaftGroupEventsListener besides counters logic removal and "zone." prefixed keys usage?

Reply:
`ZoneRebalanceRaftGroupEventsListener` also doesn't have counters logic

Reply:
> besides counters

> also doesn't have counters logic

Seems that it's the same.